clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -598,16 +607,28 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
             } else if (transactionManager.isTransactional()) {
                 transactionManager.transitionToAbortableError(exception);
             }
+            transactionManager.stopTrackingBatch(batch);
+            if (adjustSequenceNumbers)
+                transactionManager.adjustSequencesDueToFailedBatch(batch);
         }
+
+        this.sensors.recordErrors(batch.topicPartition.topic(), batch.recordCount);
         batch.done(baseOffset, logAppendTime, exception);
         this.accumulator.deallocate(batch);
     }
 
     /**
-     * We can retry a send if the error is transient and the number of attempts taken is fewer than the maximum allowed
+     * We can retry a send if the error is transient and the number of attempts taken is fewer than the maximum allowed.
+     * We can also retry OutOfOrderSequence exceptions for future batches, since if the first batch has failed, the future
+     * batches are certain to fail with an OutOfOrderSequence exception.
      */
     private boolean canRetry(ProducerBatch batch, Errors error) {
-        return batch.attempts() < this.retries && error.exception() instanceof RetriableException;
+        return batch.attempts() < this.retries &&
+                ((error.exception() instanceof RetriableException) ||
+                        (error.exception() instanceof OutOfOrderSequenceException &&
+                                transactionManager.hasProducerId(batch.producerId()) &&
```

Comment:
nit: a little difficult to follow these checks. I wonder if we could wrap them in a `TransactionManager` method with a friendly name?